README.md
<h2><strong>PUFFScoin</strong></h2>
<p>More information at puffscoin.com or visit our ANN thread at https://bitcointalk.org/index.php?topic=**TODO**</p>
<h3><strong>Coin specs:</strong></h3>
<p><strong><span style="color: #800080;"><em>Algo:</em></span></strong></p>
<ul>
<li>Xevan</li>
</ul>
<p><strong><span style="color: #800080;"><em>PoW Block Reward:</em></span></strong></p>
<ul>
    <li>[block# 1] GENESIS BLOCK - 100 mln PUFFS. Detailed statement below. </li>
    <li>[block# 2-2880] 20% MN (0.4 PUFFS) / 80% MINER (1.6 PUFFS)</li> 
    <li> [block# 2881-5760] 20% MN (0.4 PUFFS) / 70% MINER (1.4 PUFFS) / 10% Available to the Veteran's Endowment Fund (0.2 PUFFS)</li> 
    <li> [block# 5761-10080] 45% MN (0.9/n PUFFS) / 45% MINER (0.9/n PUFFS) / 10% Available to the Veteran's Endowment Fund (0.2 PUFFS)</li>
    <p> Last Proof of Work block: 10080 (approximately one week)</p>
    <p> We will have a very limited POW phase with low rewards beyond the Genesis Block fund. We will be using this period to ensure masternodes are operating efficiently, transaction obfuscation and instant send functions are operating, and so on.</p>
    <p> You can view our Genesis Sale schedule, including our purchase agenda for each phase and the escrowed checkpoints we will satisfy along the way, along with the PUFFScoin Genesis Sale terms and conditions at www.puffscoin.com/genesis.html.</p>
</ul>
<p><strong><span style="color: #800080;"><em>PoS Block Reward:</em></span></strong></p>
<ul>
<li>[block# 10081-Infinite] Fluctuates based on SeeSaw Reward Mechanism (Masternode:stakeholder saturation based) and a specific Block Reward schedule during the first two years of blockchain operation based on the Mandelbrot Curve. This reward structure will ensure that the protocol will initiate the minting of PUFFScoin to a balance essentially equal to the simple staking Cauldron and Developer Reserves. Our goal is to reach equilibrium between MasterNode ownership and simple stakeholding wallets to create a 20%:15% APR respectively. Our hope is that this coin generation agenda will encourage early MasterNode investment while ROI turnaround is very rapid, and maintain this sort of investment opportunity over the first two years of operation when the network reward levels off to 20 PUFFS/block.</li>
</ul>
<p><strong><span style="color: #800080;"><em>Block Time:</em></span></strong></p>
<ul>
<li>60 Seconds with Difficulty Retargeting every Block</li>
</ul>
<p><strong><span style="color: #800080;"><em>Max Coin Supply:</em></span></strong></p>
<ul>
<li>PoW Phase: approximately 1 week. low reward. for administrative purposes. 20,160 PUFFS generated. These will be donated to build four MasterNodes that will be raffled off during our first quarter of operation through social media campaign participation.</li>
</ul>
<p><strong><span style="color: #800080;"><em>Max Coin Supply:</em></span></strong></p>
<ul>
<li>PoS Phase: Infinte, although we may revisit the minting schedule for natural degradation of coin generation through structured epochs.</li>
</ul>
<p><strong><span style="color: #800080;"><em>PoW Phase: essentially private</em></span></strong></p>
<ul>
<li>[block# 1] 100,000,000 PUFFS Premine to create Cauldron active-transaction wallet for dispensary (50 mln), Genesis Sale funds (5 mln, 50% scheduled for the Sale and 50% set aside if demand from community demands it, otherwise these funds will be set towards bounties and social media campaigns following the Genesis Sale), 4 equal-weight developers reserves (10 mln each), as well as an Early Bounties Fund (2.5 mln) and a Masternode builders wallet and contingencies (2.5 mln). We will be giving away several Masternodes in our first few months of blockchain operation, as well as constructing a minimum of 10 Masternodes (50,000 PUFFS) administered by the Leafy Cauldron Apothercary, with profits going into the Veterans Endowment Fund. </li>
</ul>
<p><strong><span style="color: #800080;"><em>PoS Rewards Breakdown:</em></span></strong></p>
<ul>
    <p> (90% distributed to staker and masternode - 10% available to Veteran's Endowment Fund)</p>
<li>PoS Phase 01: [blocks 10081-20160]   (1 week)  3 PUFFS    </li>
<li>PoS Phase 02: [blocks 20161-40320]   (2 weeks) 5 PUFFS    </li>
<li>PoS Phase 03: [blocks 40321-60480]   (2 weeks) 10 PUFFS   </li>
<li>PoS Phase 04: [blocks 60481-80640]   (2 weeks) 15 PUFFS   </li>
<li>PoS Phase 05: [blocks 80641-100800]  (2 weeks) 22.5 PUFFS </li>
<li>PoS Phase 06: [blocks 100801-120960] (2 weeks) 30 PUFFS   </li>
<li>PoS Phase 07: [blocks 120961-141120] (2 weeks) 40 PUFFS   </li>
<li>PoS Phase 08: [blocks 141121-161280] (2 weeks) 50 PUFFS   </li>
<li>PoS Phase 09: [blocks 161281-181440] (2 weeks) 62.5 PUFFS </li>
<li>PoS Phase 10: [blocks 181441-201600] (2 weeks) 75 PUFFS   </li>    
<li>PoS Phase 11: [blocks 201661-221760] (2 weeks) 90 PUFFS   </li>  
<li>PoS Phase 12: [blocks 221761-241920] (2 weeks) 105 PUFFS  </li>
<li>PoS Phase 13: [blocks 241921-262080] (2 weeks) 122.5 PUFFS</li>
<li>PoS Phase 14: [blocks 262081-282240] (2 weeks) 140 PUFFS  </li>
<li>PoS Phase 15: [blocks 282241-302400] (2 weeks) 160 PUFFS  </li>
<li>PoS Phase 16: [blocks 302401-322560] (2 weeks) 180 PUFFS  </li>
<li>PoS Phase 17: [blocks 322561-342720] (2 weeks) 200 PUFFS  </li>
<li>PoS Phase 18: [blocks 342721-362880] (2 weeks) 220 PUFFS  </li>
<li>PoS Phase 19: [blocks 362881-383040] (2 weeks) 230 PUFFS  </li>
<li>PoS Phase 20: [blocks 383041-403200] (2 weeks) 240 PUFFS  </li>
<li>PoS Phase 21: [blocks 403201-483840] (8 weeks) 250 PUFFS  </li>
<li>PoS Phase 22: [blocks 483841-524160] (4 weeks) 200 PUFFS  </li>    
<li>PoS Phase 23: [blocks 524161-564480] (4 weeks) 150 PUFFS  </li>
<li>PoS Phase 24: [blocks 564481-604800] (4 weeks) 100 PUFFS  </li>
<li>PoS Phase 25: [blocks 604801-645120] (4 weeks) 87.5 PUFFS </li>
<li>PoS Phase 26: [blocks 645121-685440] (4 weeks) 75 PUFFS   </li>
<li>PoS Phase 27: [blocks 685441-725760] (4 weeks) 62.5 PUFFS </li>
<li>PoS Phase 28: [blocks 725761-766080] (4 weeks) 50 PUFFS   </li>
<li>PoS Phase 29: [blocks 766081-806400] (4 weeks) 45 PUFFS   </li>
<li>PoS Phase 30: [blocks 806401-846720] (4 weeks) 40 PUFFS   </li>
<li>PoS Phase 31: [blocks 846721-887040] (4 weeks) 35 PUFFS   </li>
<li>PoS Phase 32: [blocks 887041-927360] (4 weeks) 30 PUFFS   </li>
<li>PoS Phase 33: [blocks 927361-967680] (4 weeks) 27.5 PUFFS </li>
<li>PoS Phase 34: [blocks 967681-1008000](4 weeks) 25 PUFFS   </li>
<li>PoS Phase 35: [blocks 1008001-1048320](4 weeks)22.5 PUFFS </li>
<li>PoS Phase 36: [blocks 1048321-ongoing] 20 PUFFS</li>
</ul>
<br/>
<p>&nbsp;</p>
<p>+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++</p>
<p>&nbsp;</p>

doc/release-notes.md
PuffsCoin Core version 1.0.0 is now available from:

  <https://github.com/SpacemanSpliff99/PUFFSco./releases>

This is our Genesis version release, based upon PIVX 2.2.1.

Please report bugs using the issue tracker at github:

  <https://github.com/SpacemanSpliff99/PUFFSco./issues>

We will investigate and upgrade our PUFFScoin Core with updates, bug fixes, and improvements implemented through PIVX 3.0.0 throughout calendar year 2018.

Compatibility
==============

PUFFScoin Core is extensively tested on multiple operating systems using the Linux kernel, macOS 10.8+, and Windows Vista and later.

Microsoft ended support for Windows XP on [April 8th, 2014](https://www.microsoft.com/en-us/WindowsForBusiness/end-of-xp-support),
No attempt is made to prevent installing or running the software on Windows XP, you can still do so at your own risk but be aware that there are known instabilities and issues. Please do not report issues about Windows XP to the issue tracker.

PUFFScoin Core should also work on most other Unix-like systems but is not frequently tested on them.

Notable from the PIVX 2.2.1 release:

- A new command (`getstakingstatus`) has been added that returns the internal conditions
  for staking to be activated and their status.

- KeePass integration has been removed for the time being due to various inefficiencies
  with it's code.


2.2.1 Change log
=================

Detailed release notes follow. This overview includes changes that affect
behavior, not code moves, refactors and string updates. For convenience in locating
the code changes and accompanying discussion, both the pull request and
git merge commit are mentioned.

### RPC and other APIs
- #130 `ccb1526` [RPC] Add `getstakingstatus` method
- #138 `4319af3` [RPC] Require password when using UnlockAnonymizeOnly
- #142 `6b5cf7f` [RPC] Remove Keepass code due to Valgrind warnings

### Block and Transaction Handling
- #146 `bce67cb` [Wallet] Look at last CoinsView block for corruption fix process
- #154 `1b3c0d7` [Consensus] Don't pass the genesis block through CheckWork

### P2P Protocol and Network Code
- #168 `ac912d9` [Wallet] Update checkpoints with v2.2 chain
- #162 `0c0d080` Remove legacy Dash code IsReferenceNode
- #163 `96b8b00` [P2P] Change alert key to effectively disable it

### GUI
- #131 `238977b` [Qt] Adds base CSS styles for various elements
- #134 `f7cabbe` [Qt] Edit masternode.conf in Qt-wallet
- #135 `f8f1904` [Qt] Show path to wallet.dat in wallet-repair tab
- #136 `53705f1` [Qt] Fix false flags for MultiSend notification when sending transactions
- #137 `ad08051` [Qt] Fix Overview Page Balances when receiving
- #141 `17a9e0f` [Qt] Squashed trading removal code
- #151 `0409b12` [Qt] Avoid OpenSSL certstore-related memory leak
- #165 `0dad320` [Qt] More place for long locales

### Miscellaneous
- #133 `fceb421` [Docs] Add GitHub Issue template and Contributor guidelines
- #144 `e4e68bc` [Wallet] Reduce usage of atoi to comply with CWE-190
- #152 `6a1de07` [Trivial] Use LogPrint for repetitive budget logs
- #157 `41fdeaa` [Budget] Add log for removed budget proposals
- #166 `d37b4aa` [Utils] Add ExecStop= to example systemd service
- #167 `a6becee` [Utils] makeseeds script update

Credits via PIVX 2.2.1
======================

Thanks to everyone who directly contributed to this release:

- Aaron Miller
- Fuzzbawls
- Mrs-X
- Spock
- presstab


src/hash.h
// Copyright (c) 2009-2010 Satoshi Nakamoto
// Copyright (c) 2009-2014 The Bitcoin developers
// Copyright (c) 2014-2015 The Dash developers
// Copyright (c) 2015-2017 The AmsterdamCoin developers
// Copyright (c) 2017 The PuffsCoin developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#ifndef XEVAN_H
#define XEVAN_H

#include "crypto/ripemd160.h"
#include "crypto/sha256.h"
#include "prevector.h"
#include "serialize.h"
#include "uint256.h"
#include "version.h"

#include "crypto/sph_blake.h"
#include "crypto/sph_bmw.h"
#include "crypto/sph_groestl.h"
#include "crypto/sph_jh.h"
#include "crypto/sph_keccak.h"
#include "crypto/sph_skein.h"
#include "crypto/sph_luffa.h"
#include "crypto/sph_cubehash.h" 
#include "crypto/sph_shavite.h"
#include "crypto/sph_simd.h"
#include "crypto/sph_echo.h" 
#include "crypto/sph_hamsi.h"  
#include "crypto/sph_fugue.h" 
#include "crypto/sph_shabal.h"  
#include "crypto/sph_whirlpool.h"  
#include "crypto/sph_sha2.h"  
#include "crypto/sph_haval.h"  

#include <iomanip>
#include <openssl/sha.h>
#include <sstream>
#include <vector>
using namespace std;
typedef uint256 ChainCode;

#ifdef GLOBALDEFINED
#define GLOBAL
#else
#define GLOBAL extern
#endif

GLOBAL sph_blake512_context     z_blake;
GLOBAL sph_bmw512_context       z_bmw;
GLOBAL sph_groestl512_context   z_groestl;
GLOBAL sph_jh512_context        z_jh;
GLOBAL sph_keccak512_context    z_keccak;
GLOBAL sph_skein512_context     z_skein;
GLOBAL sph_luffa512_context     z_luffa;
GLOBAL sph_cubehash512_context  z_cubehash;
GLOBAL sph_shavite512_context   z_shavite;
GLOBAL sph_simd512_context      z_simd;
GLOBAL sph_echo512_context      z_echo;
GLOBAL sph_hamsi512_context     z_hamsi;
GLOBAL sph_fugue512_context     z_fugue;
GLOBAL sph_shabal512_context    z_shabal;
GLOBAL sph_whirlpool_context    z_whirlpool;
GLOBAL sph_sha512_context       z_sha2;
GLOBAL sph_haval256_5_context   z_haval;

#define fillz() do { \
    sph_blake512_init(&z_blake); \
    sph_bmw512_init(&z_bmw); \
    sph_groestl512_init(&z_groestl); \
    sph_jh512_init(&z_jh); \
    sph_keccak512_init(&z_keccak); \
    sph_skein512_init(&z_skein); \
    sph_luffa512_init(&z_luffa); \
    sph_cubehash512_init(&z_cubehash); \
    sph_shavite512_init(&z_shavite); \
    sph_simd512_init(&z_simd); \
    sph_echo512_init(&z_echo); \
    sph_hamsi512_init(&z_hamsi); \
    sph_fugue512_init(&z_fugue); \
    sph_shabal512_init(&z_shabal); \
    sph_whirlpool_init(&z_whirlpool); \
    sph_sha512_init(&z_sha2); \
    sph_haval256_5_init(&z_haval); \
} while (0) 

#define ZBLAKE (memcpy(&ctx_blake, &z_blake, sizeof(z_blake)))
#define ZBMW (memcpy(&ctx_bmw, &z_bmw, sizeof(z_bmw)))
#define ZGROESTL (memcpy(&ctx_groestl, &z_groestl, sizeof(z_groestl)))
#define ZJH (memcpy(&ctx_jh, &z_jh, sizeof(z_jh)))
#define ZKECCAK (memcpy(&ctx_keccak, &z_keccak, sizeof(z_keccak)))
#define ZSKEIN (memcpy(&ctx_skein, &z_skein, sizeof(z_skein)))
#define ZHAMSI (memcpy(&ctx_hamsi, &z_hamsi, sizeof(z_hamsi)))
#define ZFUGUE (memcpy(&ctx_fugue, &z_fugue, sizeof(z_fugue)))
#define ZSHABAL (memcpy(&ctx_shabal, &z_shabal, sizeof(z_shabal)))
#define ZWHIRLPOOL (memcpy(&ctx_whirlpool, &z_whirlpool, sizeof(z_whirlpool)))
#define ZSHA2 (memcpy(&ctx_sha2, &z_sha2, sizeof(z_sha2)))
#define ZHAVAL (memcpy(&ctx_haval, &z_haval, sizeof(z_haval)))

/** A hasher class for Bitcoin's 256-bit hash (double SHA-256). */
class CHash256
{
private:
    CSHA256 sha;

public:
    static const size_t OUTPUT_SIZE = CSHA256::OUTPUT_SIZE;

    void Finalize(unsigned char hash[OUTPUT_SIZE])
    {
        unsigned char buf[sha.OUTPUT_SIZE];
        sha.Finalize(buf);
        sha.Reset().Write(buf, sha.OUTPUT_SIZE).Finalize(hash);
    }

    CHash256& Write(const unsigned char* data, size_t len)
    {
        sha.Write(data, len);
        return *this;
    }

    CHash256& Reset()
    {
        sha.Reset();
        return *this;
    }
};

/** A hasher class for Bitcoin's 160-bit hash (SHA-256 + RIPEMD-160). */
class CHash160
{
private:
    CSHA256 sha;

public:
    static const size_t OUTPUT_SIZE = CRIPEMD160::OUTPUT_SIZE;

    void Finalize(unsigned char hash[OUTPUT_SIZE])
    {
        unsigned char buf[sha.OUTPUT_SIZE];
        sha.Finalize(buf);
        CRIPEMD160().Write(buf, sha.OUTPUT_SIZE).Finalize(hash);
    }

    CHash160& Write(const unsigned char* data, size_t len)
    {
        sha.Write(data, len);
        return *this;
    }

    CHash160& Reset()
    {
        sha.Reset();
        return *this;
    }
};

inline void Hash(void* in, unsigned int len, unsigned char* out)
{
    SHA256_CTX sha256;
    SHA256_Init(&sha256);
    SHA256_Update(&sha256, in, len);
    SHA256_Final(out, &sha256);
}

/** Compute the 256-bit hash of an object. */
template <typename T1>
inline uint256 Hash(const T1 pbegin, const T1 pend)
{
    static const unsigned char pblank[1] = {};
    uint256 result;
    CHash256().Write(pbegin == pend ? pblank : (const unsigned char*)&pbegin[0], (pend - pbegin) * sizeof(pbegin[0])).Finalize((unsigned char*)&result);
    return result;
}

/** Compute the 256-bit hash of the concatenation of two objects. */
template <typename T1, typename T2>
inline uint256 Hash(const T1 p1begin, const T1 p1end, const T2 p2begin, const T2 p2end)
{
    static const unsigned char pblank[1] = {};
    uint256 result;
    CHash256().Write(p1begin == p1end ? pblank : (const unsigned char*)&p1begin[0], (p1end - p1begin) * sizeof(p1begin[0])).Write(p2begin == p2end ? pblank : (const unsigned char*)&p2begin[0], (p2end - p2begin) * sizeof(p2begin[0])).Finalize((unsigned char*)&result);
    return result;
}

/** Compute the 256-bit hash of the concatenation of three objects. */
template <typename T1, typename T2, typename T3>
inline uint256 Hash(const T1 p1begin, const T1 p1end, const T2 p2begin, const T2 p2end, const T3 p3begin, const T3 p3end)
{
    static const unsigned char pblank[1] = {};
    uint256 result;
    CHash256().Write(p1begin == p1end ? pblank : (const unsigned char*)&p1begin[0], (p1end - p1begin) * sizeof(p1begin[0])).Write(p2begin == p2end ? pblank : (const unsigned char*)&p2begin[0], (p2end - p2begin) * sizeof(p2begin[0])).Write(p3begin == p3end ? pblank : (const unsigned char*)&p3begin[0], (p3end - p3begin) * sizeof(p3begin[0])).Finalize((unsigned char*)&result);
    return result;
}

/** Compute the 256-bit hash of the concatenation of three objects. */
template <typename T1, typename T2, typename T3, typename T4>
inline uint256 Hash(const T1 p1begin, const T1 p1end, const T2 p2begin, const T2 p2end, const T3 p3begin, const T3 p3end, const T4 p4begin, const T4 p4end)
{
    static const unsigned char pblank[1] = {};
    uint256 result;
    CHash256().Write(p1begin == p1end ? pblank : (const unsigned char*)&p1begin[0], (p1end - p1begin) * sizeof(p1begin[0])).Write(p2begin == p2end ? pblank : (const unsigned char*)&p2begin[0], (p2end - p2begin) * sizeof(p2begin[0])).Write(p3begin == p3end ? pblank : (const unsigned char*)&p3begin[0], (p3end - p3begin) * sizeof(p3begin[0])).Write(p4begin == p4end ? pblank : (const unsigned char*)&p4begin[0], (p4end - p4begin) * sizeof(p4begin[0])).Finalize((unsigned char*)&result);
    return result;
}

/** Compute the 256-bit hash of the concatenation of three objects. */
template <typename T1, typename T2, typename T3, typename T4, typename T5>
inline uint256 Hash(const T1 p1begin, const T1 p1end, const T2 p2begin, const T2 p2end, const T3 p3begin, const T3 p3end, const T4 p4begin, const T4 p4end, const T5 p5begin, const T5 p5end)
{
    static const unsigned char pblank[1] = {};
    uint256 result;
    CHash256().Write(p1begin == p1end ? pblank : (const unsigned char*)&p1begin[0], (p1end - p1begin) * sizeof(p1begin[0])).Write(p2begin == p2end ? pblank : (const unsigned char*)&p2begin[0], (p2end - p2begin) * sizeof(p2begin[0])).Write(p3begin == p3end ? pblank : (const unsigned char*)&p3begin[0], (p3end - p3begin) * sizeof(p3begin[0])).Write(p4begin == p4end ? pblank : (const unsigned char*)&p4begin[0], (p4end - p4begin) * sizeof(p4begin[0])).Write(p5begin == p5end ? pblank : (const unsigned char*)&p5begin[0], (p5end - p5begin) * sizeof(p5begin[0])).Finalize((unsigned char*)&result);
    return result;
}

/** Compute the 256-bit hash of the concatenation of three objects. */
template <typename T1, typename T2, typename T3, typename T4, typename T5, typename T6>
inline uint256 Hash(const T1 p1begin, const T1 p1end, const T2 p2begin, const T2 p2end, const T3 p3begin, const T3 p3end, const T4 p4begin, const T4 p4end, const T5 p5begin, const T5 p5end, const T6 p6begin, const T6 p6end)
{
    static const unsigned char pblank[1] = {};
    uint256 result;
    CHash256().Write(p1begin == p1end ? pblank : (const unsigned char*)&p1begin[0], (p1end - p1begin) * sizeof(p1begin[0])).Write(p2begin == p2end ? pblank : (const unsigned char*)&p2begin[0], (p2end - p2begin) * sizeof(p2begin[0])).Write(p3begin == p3end ? pblank : (const unsigned char*)&p3begin[0], (p3end - p3begin) * sizeof(p3begin[0])).Write(p4begin == p4end ? pblank : (const unsigned char*)&p4begin[0], (p4end - p4begin) * sizeof(p4begin[0])).Write(p5begin == p5end ? pblank : (const unsigned char*)&p5begin[0], (p5end - p5begin) * sizeof(p5begin[0])).Write(p6begin == p6end ? pblank : (const unsigned char*)&p6begin[0], (p6end - p6begin) * sizeof(p6begin[0])).Finalize((unsigned char*)&result);
    return result;
}

/** Compute the 160-bit hash an object. */
template <typename T1>
inline uint160 Hash160(const T1 pbegin, const T1 pend)
{
    static unsigned char pblank[1] = {};
    uint160 result;
    CHash160().Write(pbegin == pend ? pblank : (const unsigned char*)&pbegin[0], (pend - pbegin) * sizeof(pbegin[0])).Finalize((unsigned char*)&result);
    return result;
}

/** Compute the 160-bit hash of a vector. */
inline uint160 Hash160(const std::vector<unsigned char>& vch)
{
    return Hash160(vch.begin(), vch.end());
}

/** A writer stream (for serialization) that computes a 256-bit hash. */
class CHashWriter
{
private:
    CHash256 ctx;

public:
    int nType;
    int nVersion;

    CHashWriter(int nTypeIn, int nVersionIn) : nType(nTypeIn), nVersion(nVersionIn) {}

    CHashWriter& write(const char* pch, size_t size)
    {
        ctx.Write((const unsigned char*)pch, size);
        return (*this);
    }

    // invalidates the object
    uint256 GetHash()
    {
        uint256 result;
        ctx.Finalize((unsigned char*)&result);
        return result;
    }

    template <typename T>
    CHashWriter& operator<<(const T& obj)
    {
        // Serialize to this stream
        ::Serialize(*this, obj, nType, nVersion);
        return (*this);
    }
};

/** Compute the 256-bit hash of an object's serialization. */
template <typename T>
uint256 SerializeHash(const T& obj, int nType = SER_GETHASH, int nVersion = PROTOCOL_VERSION)
{
    CHashWriter ss(nType, nVersion);
    ss << obj;
    return ss.GetHash();
}

unsigned int MurmurHash3(unsigned int nHashSeed, const std::vector<unsigned char>& vDataToHash);

void BIP32Hash(const unsigned char chainCode[32], unsigned int nChild, unsigned char header, const unsigned char data[32], unsigned char output[64]);

//int HMAC_SHA512_Init(HMAC_SHA512_CTX *pctx, const void *pkey, size_t len);
//int HMAC_SHA512_Update(HMAC_SHA512_CTX *pctx, const void *pdata, size_t len);
//int HMAC_SHA512_Final(unsigned char *pmd, HMAC_SHA512_CTX *pctx);

template<typename T1>
inline uint256 XEVAN(const T1 pbegin, const T1 pend)
{
    //LogPrintf("X11 Hash \n");
	sph_blake512_context      ctx_blake;
    sph_bmw512_context        ctx_bmw;
    sph_groestl512_context    ctx_groestl;
    sph_jh512_context         ctx_jh;
    sph_keccak512_context     ctx_keccak;
    sph_skein512_context      ctx_skein;
    sph_luffa512_context      ctx_luffa;
    sph_cubehash512_context   ctx_cubehash;
    sph_shavite512_context    ctx_shavite;
    sph_simd512_context       ctx_simd;
    sph_echo512_context       ctx_echo;
    sph_hamsi512_context      ctx_hamsi;
    sph_fugue512_context      ctx_fugue;
    sph_shabal512_context     ctx_shabal;
    sph_whirlpool_context     ctx_whirlpool;
    sph_sha512_context        ctx_sha2;
    sph_haval256_5_context    ctx_haval;
static unsigned char pblank[1];

#ifndef QT_NO_DEBUG
    //std::string strhash;
    //strhash = "";
#endif
    int worknumber =128;
    uint512 hash[34];

    sph_blake512_init(&ctx_blake);
    sph_blake512 (&ctx_blake, (pbegin == pend ? pblank : static_cast<const void*>(&pbegin[0])), (pend - pbegin) * sizeof(pbegin[0]));
    sph_blake512_close(&ctx_blake, static_cast<void*>(&hash[0]));
    
    sph_bmw512_init(&ctx_bmw);
    sph_bmw512 (&ctx_bmw, static_cast<const void*>(&hash[0]), worknumber);
    sph_bmw512_close(&ctx_bmw, static_cast<void*>(&hash[1]));

    sph_groestl512_init(&ctx_groestl);
    sph_groestl512 (&ctx_groestl, static_cast<const void*>(&hash[1]), worknumber);
    sph_groestl512_close(&ctx_groestl, static_cast<void*>(&hash[2]));

    sph_skein512_init(&ctx_skein);
    sph_skein512 (&ctx_skein, static_cast<const void*>(&hash[2]), worknumber);
    sph_skein512_close(&ctx_skein, static_cast<void*>(&hash[3]));
    
    sph_jh512_init(&ctx_jh);
    sph_jh512 (&ctx_jh, static_cast<const void*>(&hash[3]), worknumber);
    sph_jh512_close(&ctx_jh, static_cast<void*>(&hash[4]));
    
    sph_keccak512_init(&ctx_keccak);
    sph_keccak512 (&ctx_keccak, static_cast<const void*>(&hash[4]), worknumber);
    sph_keccak512_close(&ctx_keccak, static_cast<void*>(&hash[5]));

    sph_luffa512_init(&ctx_luffa);
    sph_luffa512 (&ctx_luffa, static_cast<void*>(&hash[5]), worknumber);
    sph_luffa512_close(&ctx_luffa, static_cast<void*>(&hash[6]));
    
    sph_cubehash512_init(&ctx_cubehash);
    sph_cubehash512 (&ctx_cubehash, static_cast<const void*>(&hash[6]), worknumber);
    sph_cubehash512_close(&ctx_cubehash, static_cast<void*>(&hash[7]));
    
    sph_shavite512_init(&ctx_shavite);
    sph_shavite512(&ctx_shavite, static_cast<const void*>(&hash[7]), worknumber);
    sph_shavite512_close(&ctx_shavite, static_cast<void*>(&hash[8]));
        
    sph_simd512_init(&ctx_simd);
    sph_simd512 (&ctx_simd, static_cast<const void*>(&hash[8]), worknumber);
    sph_simd512_close(&ctx_simd, static_cast<void*>(&hash[9]));

    sph_echo512_init(&ctx_echo);
    sph_echo512 (&ctx_echo, static_cast<const void*>(&hash[9]), worknumber);
    sph_echo512_close(&ctx_echo, static_cast<void*>(&hash[10]));

    sph_hamsi512_init(&ctx_hamsi);
    sph_hamsi512 (&ctx_hamsi, static_cast<const void*>(&hash[10]), worknumber);
    sph_hamsi512_close(&ctx_hamsi, static_cast<void*>(&hash[11]));

    sph_fugue512_init(&ctx_fugue);
    sph_fugue512 (&ctx_fugue, static_cast<const void*>(&hash[11]), worknumber);
    sph_fugue512_close(&ctx_fugue, static_cast<void*>(&hash[12]));

    sph_shabal512_init(&ctx_shabal);
    sph_shabal512 (&ctx_shabal, static_cast<const void*>(&hash[12]), worknumber);
    sph_shabal512_close(&ctx_shabal, static_cast<void*>(&hash[13]));

    sph_whirlpool_init(&ctx_whirlpool);
    sph_whirlpool (&ctx_whirlpool, static_cast<const void*>(&hash[13]), worknumber);
    sph_whirlpool_close(&ctx_whirlpool, static_cast<void*>(&hash[14]));

    sph_sha512_init(&ctx_sha2);
    sph_sha512 (&ctx_sha2, static_cast<const void*>(&hash[14]), worknumber);
    sph_sha512_close(&ctx_sha2, static_cast<void*>(&hash[15]));

    sph_haval256_5_init(&ctx_haval);
    sph_haval256_5 (&ctx_haval, static_cast<const void*>(&hash[15]), worknumber);
    sph_haval256_5_close(&ctx_haval, static_cast<void*>(&hash[16]));
	
    ///  Part2
    sph_blake512_init(&ctx_blake);
    sph_blake512 (&ctx_blake, static_cast<const void*>(&hash[16]), worknumber);
    sph_blake512_close(&ctx_blake, static_cast<void*>(&hash[17]));
    
    sph_bmw512_init(&ctx_bmw);
    sph_bmw512 (&ctx_bmw, static_cast<const void*>(&hash[17]), worknumber);
    sph_bmw512_close(&ctx_bmw, static_cast<void*>(&hash[18]));

    sph_groestl512_init(&ctx_groestl);
    sph_groestl512 (&ctx_groestl, static_cast<const void*>(&hash[18]), worknumber);
    sph_groestl512_close(&ctx_groestl, static_cast<void*>(&hash[19]));

    sph_skein512_init(&ctx_skein);
    sph_skein512 (&ctx_skein, static_cast<const void*>(&hash[19]), worknumber);
    sph_skein512_close(&ctx_skein, static_cast<void*>(&hash[20]));
    
    sph_jh512_init(&ctx_jh);
    sph_jh512 (&ctx_jh, static_cast<const void*>(&hash[20]), worknumber);
    sph_jh512_close(&ctx_jh, static_cast<void*>(&hash[21]));
    
    sph_keccak512_init(&ctx_keccak);
    sph_keccak512 (&ctx_keccak, static_cast<const void*>(&hash[21]), worknumber);
    sph_keccak512_close(&ctx_keccak, static_cast<void*>(&hash[22]));

    sph_luffa512_init(&ctx_luffa);
    sph_luffa512 (&ctx_luffa, static_cast<void*>(&hash[22]), worknumber);
    sph_luffa512_close(&ctx_luffa, static_cast<void*>(&hash[23]));
    
    sph_cubehash512_init(&ctx_cubehash);
    sph_cubehash512 (&ctx_cubehash, static_cast<const void*>(&hash[23]), worknumber);
    sph_cubehash512_close(&ctx_cubehash, static_cast<void*>(&hash[24]));
    
    sph_shavite512_init(&ctx_shavite);
    sph_shavite512(&ctx_shavite, static_cast<const void*>(&hash[24]), worknumber);
    sph_shavite512_close(&ctx_shavite, static_cast<void*>(&hash[25]));
        
    sph_simd512_init(&ctx_simd);
    sph_simd512 (&ctx_simd, static_cast<const void*>(&hash[25]), worknumber);
    sph_simd512_close(&ctx_simd, static_cast<void*>(&hash[26]));

    sph_echo512_init(&ctx_echo);
    sph_echo512 (&ctx_echo, static_cast<const void*>(&hash[26]), worknumber);
    sph_echo512_close(&ctx_echo, static_cast<void*>(&hash[27]));

    sph_hamsi512_init(&ctx_hamsi);
    sph_hamsi512 (&ctx_hamsi, static_cast<const void*>(&hash[27]), worknumber);
    sph_hamsi512_close(&ctx_hamsi, static_cast<void*>(&hash[28]));

    sph_fugue512_init(&ctx_fugue);
    sph_fugue512 (&ctx_fugue, static_cast<const void*>(&hash[28]), worknumber);
    sph_fugue512_close(&ctx_fugue, static_cast<void*>(&hash[29]));

    sph_shabal512_init(&ctx_shabal);
    sph_shabal512 (&ctx_shabal, static_cast<const void*>(&hash[29]), worknumber);
    sph_shabal512_close(&ctx_shabal, static_cast<void*>(&hash[30]));

    sph_whirlpool_init(&ctx_whirlpool);
    sph_whirlpool (&ctx_whirlpool, static_cast<const void*>(&hash[30]), worknumber);
    sph_whirlpool_close(&ctx_whirlpool, static_cast<void*>(&hash[31]));

    sph_sha512_init(&ctx_sha2);
    sph_sha512 (&ctx_sha2, static_cast<const void*>(&hash[31]), worknumber);
    sph_sha512_close(&ctx_sha2, static_cast<void*>(&hash[32]));

    sph_haval256_5_init(&ctx_haval);
    sph_haval256_5 (&ctx_haval, static_cast<const void*>(&hash[32]), worknumber);
    sph_haval256_5_close(&ctx_haval, static_cast<void*>(&hash[33]));


    return hash[33].trim256();
}

void scrypt_hash(const char* pass, unsigned int pLen, const char* salt, unsigned int sLen, char* output, unsigned int N, unsigned int r, unsigned int p, unsigned int dkLen);

#endif // BITCOIN_HASH_H
